File: core/src/main/java/at/shockbytes/corey/common/core/workout/model/Workout.java
package at.shockbytes.corey.common.core.workout.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import at.shockbytes.corey.common.core.R;


/**
 * @author Martin Macheiner
 *         Date: 27.10.2015.
 */
public class Workout implements Comparable<Workout>, Parcelable {

    public enum Intensity {EASY, MEDIUM, HARD, BEAST}

    public enum BodyRegion {LEGS, CORE, ARMS, CHEST, FULL_BODY}

    private String id;

    private String name;
    private int duration; // in minutes
    private List<Exercise> exercises;

    private Intensity intensity;
    private BodyRegion bodyRegion;

    public Workout() {
        this("", 0, Intensity.MEDIUM, BodyRegion.CORE);
    }

    public Workout(String name, int duration, Intensity intensity, BodyRegion bodyRegion) {
        this(name, duration, intensity, bodyRegion, new ArrayList<Exercise>());
    }

    public Workout(String name, int duration, Intensity intensity,
                   BodyRegion bodyRegion, ArrayList<Exercise> exercises) {
        setName(name);
        this.duration = duration;
        this.intensity = intensity;
        this.bodyRegion = bodyRegion;
        this.exercises = exercises;
    }

    protected Workout(Parcel in) {
        id = in.readString();
        name = in.readString();
        duration = in.readInt();
        bodyRegion = BodyRegion.values()[in.readInt()];
        intensity = Intensity.values()[in.readInt()];
        Parcelable[] parcels = in.readParcelableArray(Exercise.class.getClassLoader());
        exercises = new ArrayList<>();
        for (Parcelable p: parcels) {
            exercises.add((Exercise) p);
        }
    }

    public String getName() {
        return name;
    }

    public Workout setName(String name) {
        this.name = "\"" + name + "\"";
        return this;
    }

    public Workout setDuration(int duration) {
        this.duration = duration;
        return this;
    }

    public int getDuration() {
        return duration;
    }

    public Workout setExercises(List<Exercise> exercises) {
        this.exercises = exercises;
        return this;
    }

    public int getExerciseCount() {
        return exercises.size();
    }

    public List<Exercise> getExercises() {
        return exercises;
    }

    public Workout setIntensity(Intensity intensity) {
        this.intensity = intensity;
        return this;
    }

    public Workout setBodyRegion(BodyRegion bodyRegion) {
        this.bodyRegion = bodyRegion;
        return this;
    }

    public Intensity getIntensity() {
        return intensity;
    }

    public int getColorResForIntensity() {

        int color = 0;
        switch (intensity) {

            case EASY:
                color = R.color.workout_intensity_easy;
                break;
            case MEDIUM:
                color = R.color.workout_intensity_medium;
                break;
            case HARD:
                color = R.color.workout_intensity_hard;
                break;
            case BEAST:
                color = R.color.workout_intensity_beast;
                break;
        }
        return color;
    }

    public int getDarkColorResForIntensity() {

        int color = 0;
        switch (intensity) {

            case EASY:
                color = R.color.workout_intensity_easy_dark;
                break;
            case MEDIUM:
                color = R.color.workout_intensity_medium_dark;
                break;
            case HARD:
                color = R.color.workout_intensity_hard_dark;
                break;
            case BEAST:
                color = R.color.workout_intensity_beast_dark;
                break;
        }
        return color;
    }

    public int getImageResForBodyRegion() {

        int bodyRegionImage = 0;
        switch (bodyRegion) {

            case LEGS:
                bodyRegionImage = R.drawable.ic_bodyregion_legs;
                break;

            case CORE:
                bodyRegionImage = R.drawable.ic_bodyregion_core;
                break;

            case ARMS:
                bodyRegionImage = R.drawable.ic_bodyregion_arms;
                break;

            case CHEST:
                bodyRegionImage = R.drawable.ic_bodyregion_chest;
                break;

            case FULL_BODY:
                bodyRegionImage = R.drawable.ic_bodyregion_whole;
                break;
        }
        return bodyRegionImage;
    }

    public BodyRegion getBodyRegion() {
        return bodyRegion;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public static final Creator<Workout> CREATOR = new Creator<Workout>() {
        @Override
        public Workout createFromParcel(Parcel in) {
            return new Workout(in);
        }

        @Override
        public Workout[] newArray(int size) {
            return new Workout[size];
        }
    };

    @Override
    public int compareTo(@NonNull Workout another) {
        return id.compareTo(another.id);
    }

    @Override
    public boolean equals(Object obj) {
        return obj != null && obj instanceof Workout && compareTo((Workout) obj) == 0;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeString(id);
        parcel.writeString(name);
        parcel.writeInt(duration);
        parcel.writeInt(bodyRegion.ordinal());
        parcel.writeInt(intensity.ordinal());
        parcel.writeParcelableArray(exercises.toArray(new Exercise[exercises.size()]), i);
    }

    @Override
    public String toString() {

        String str = "";
        for (Exercise ex : exercises) {
            str += ex.toString()+"\n\n";
        }

        return "Title:\t" + name +
                "\nId:" + id +
                "\nDuration: " + duration + " min" +
                "\nIntensity: " + intensity +
                "\nMuscles: " + bodyRegion +
                "\nExercises: " + exercises.size() +
                "\n" + str;
    }
}

File: core/src/main/java/at/shockbytes/corey/common/core/workout/model/Workout.kt
package at.shockbytes.corey.common.core.workout.model

import android.os.Parcelable
import at.shockbytes.corey.common.core.R
import com.google.firebase.database.Exclude
import kotlinx.android.parcel.Parcelize

/**
 * Author:  Martin Macheiner
 * Date:    27.10.2015
 */
@Parcelize
data class Workout(
    var id: String = "",
    private var name: String = "", // private necessary for custom setter
    var duration: Int = 0, // in minutes
    var exercises: MutableList<Exercise> = mutableListOf(),
    var intensity: Intensity = Intensity.MEDIUM,
    var bodyRegion: BodyRegion = BodyRegion.CORE,
    var locationType: LocationType = LocationType.INDOOR
) : Comparable<Workout>, Parcelable {

    enum class Intensity {
        EASY, MEDIUM, HARD, BEAST
    }

    enum class BodyRegion {
        LEGS, CORE, ARMS, CHEST, FULL_BODY
    }

    val displayableName: String
        @Exclude
        get() = name.replace("_".toRegex(), " ")

    val exerciseCount: Int
        get() = exercises.size

    val colorResForIntensity: Int
        @Exclude
        get() {
            return when (intensity) {
                Workout.Intensity.EASY -> R.color.workout_intensity_easy
                Workout.Intensity.MEDIUM -> R.color.workout_intensity_medium
                Workout.Intensity.HARD -> R.color.workout_intensity_hard
                Workout.Intensity.BEAST -> R.color.workout_intensity_beast
            }
        }

    val darkColorResForIntensity: Int
        @Exclude
        get() {
            return when (intensity) {
                Workout.Intensity.EASY -> R.color.workout_intensity_easy_dark
                Workout.Intensity.MEDIUM -> R.color.workout_intensity_medium_dark
                Workout.Intensity.HARD -> R.color.workout_intensity_hard_dark
                Workout.Intensity.BEAST -> R.color.workout_intensity_beast_dark
            }
        }

    val imageResForBodyRegion: Int
        @Exclude
        get() {
            return when (bodyRegion) {
                Workout.BodyRegion.LEGS -> R.drawable.ic_bodyregion_legs
                Workout.BodyRegion.CORE -> R.drawable.ic_bodyregion_core
                Workout.BodyRegion.ARMS -> R.drawable.ic_bodyregion_arms
                Workout.BodyRegion.CHEST -> R.drawable.ic_bodyregion_chest
                Workout.BodyRegion.FULL_BODY -> R.drawable.ic_bodyregion_whole
            }
        }

    val equipment: Equipment
        @Exclude
        get() {
            val max = exercises.map { it.equipment.ordinal }.max() ?: 0
            return Equipment.fromIndex(max)
        }

    fun setName(value: String) {
        name = value.replace(" ", "_")
    }

    fun getName(): String {
        return name
    }

    override fun compareTo(other: Workout): Int {
        return id.compareTo(other.id)
    }
}
